schematic/base.py

```diff
@@ -0,0 +1,446 @@
+        return [col.name for col in self.columns]
+
+    def get_rows(self, *args, **kwargs):
+        """Generator for rows of the table described by this TableDefinition.
+
+        Args:
+          *args, **kwargs: DB-specific arguments for connection, e.g.,
+                           a psycopg2.connection or csv.DictReader object
+        Raises:
+          NotImplementedError: Subclasses should implement this.
+        """
+        raise NotImplementedError
+
+    @classmethod
+    def from_source(cls, *args, **kwargs):
+        """Instantiate from an implementation-specific source (e.g., a CSV file or a DB connection
+
+        Raises:
+          NotImplementedError: Subclasses should implement this.
+        """
+        raise NotImplementedError
+
+
+class Schematic(ABC, DictableMixin):
+    """Interface for implementation specifics for a type of database or warehouse.
+
+    The TableColumnTypes in a given schematic form a tree with the most
+    restrictive types being leaf nodes and the least restrictive type
+    being the root node.
+
+    Attributes:
+      name: Static attribute with the name of this schematic
+      most_restrictive_types: the leaf nodes of the restrictivity tree
+      table_def: implementation of TableDefinition for this schematic
+    """
+    name = 'schematic'
+    most_restrictive_types = []
+    table_definition_class = TableDefinition
+    column_class = TableColumn
+    null_strings = []
+
+    def get_distance_from_leaf_node(self, column_type):
+        """Get the distance between the given TableColumnType
+        and its nearest leaf node.
+
+        Returns:
+          An int
+        """
+        distances = []
+        for ct in self.most_restrictive_types:
+            distance = 0
+            nlr = ct
+            while nlr:
+                if nlr == column_type:
+                    distances.append(distance)
+                    nlr = None
+                else:
+                    distance += 1
+                    nlr = nlr.next_less_restrictive
+        if not distances:
+            raise ColumnTypeNotFoundError
+        return min(distances)
+
+    def get_type(self, value, previous_type=None):
+        """Get what type of column the given value would be.
+
+        Args:
+          value: the value to get the type for.
+          previous_type: the type that the column this value is in
+                         had previously been assigned
+        Returns:
+          TableColumnType that can fit the value and all values
+          of previous_type.
+        Raises:
+          ValueError: if the given value can't fit into a column
+                      of any type in this Schematic
+        """
+        if value in self.null_strings:
+            return previous_type
+        if not previous_type:
+            depth_dict = {}
+            for column_type in self.column_types():
+                if column_type()._value_is_compatible_superset(value):
```

Comment:
```suggestion
                if column_type().value_is_compatible_superset(value):
```
Make this method public if we're calling it from here